### DISCUSSION TITLE
Flashbot bundle support

### DISCUSSION THREAD

**Discussion by 0xsha** [wevm/viem#1746] [Idea / Feature Request]
hi guys, 

I think this is extremely nice to have because like IPC it allows Vim to be used to beyond web As more and more builders support Flashbots-specific RPCs, it would be super cool if there were native Vim support. For starters, it can support just the main calls (base proivder) and then move towards full support with new endpoints and calls (which are not implemented by all RPCs).

Base provider (implemented by most providers):

```
eth_sendBundle
eth_cancelBundle
eth_sendPrivateTransaction
eth_sendPrivateRawTransaction
eth_cancelPrivateTransaction
eth_callBundle
```

Extended version (implemented by fewer RPC providers):

```
mev_sendBundle
mev_simBundle
etc
```

API : https://docs.flashbots.net/flashbots-auction/advanced/rpc-endpoint
Reference Code : https://github.com/flashbots/ethers-provider-flashbots-bundle/blob/master/src/index.ts

Thanks @jxom and team for the awesome client ❤️

**Comment by sakulstra** [comment]
@jxom did you ever have a chance to look into this?

I think flashbots is no longer soso useful, but mev share is quite handy.
I just tried implementing this and the main blocker i faced are the headers required here: https://docs.flashbots.net/flashbots-auction/advanced/rpc-endpoint#authentication as i think on viem clients there is no feasible way to inject request dependent headers.

**Reply by jxom** [reply]
PRs welcome.

**Reply by sakulstra** [reply]
@jxom actually i think i managed to do it without altering the code 😅 

In retrospect my issue here was:
- imprecise documentation on https://viem.sh/docs/clients/transports/http#onfetchrequest-optional - docs say this can be used for debugging with return `void`, but factually can be used for exactly that (mutating the request).
- a bug in the error reporting of viem. https://github.com/wevm/viem/blob/main/src/utils/rpc/http.ts#L153 here the url reported will always be the "original url", even if replacing it via `onFetchRequest`


Here is the full code.
```ts
const walletClient = createWalletClient({
    account,
    chain: chain,
    transport: http(rpc, {
      onFetchRequest: async (request, init) => {
        if ((init.body as string).includes("eth_sendPrivateTransaction")) {
          const signature = await account.signMessage({
            message: keccak256(toBytes(init.body?.toString()!)),
          });
          return {
            ...init,
            url: "https://relay.flashbots.net",
            headers: {
              "X-Flashbots-Signature": `${account.address}:${signature}`,
            },
          };
        }
      },
    }),
  })
    .extend((client) => ({
      async sendPrivateTransaction(args: TransactionRequest) {
        const request = await client.prepareTransactionRequest(args);
        const signedTx = await client.signTransaction(request);
        (await client.request({
          method: "eth_sendPrivateTransaction" as any,
          params: [
            {
              tx: signedTx,
              maxBlockNumber: toHex((await getBlockNumber(client)) + 100n),
              preferences: {
                fast: true,
                privacy: {
                  hints: ["hash"],
                  builders: ["flashbots", "beaverbuild.org", "Titan", "rsync"],
                },
              },
            },
          ] as any,
        })) as SendTransactionReturnType;
        return keccak256(signedTx);
      },
    }))
```

Happy to do a pr, but honestly idk "how" to best approach this 😅
Most of these privateTxn rpcs don't allow most rpc methods, so eventually one needs to switch dependent on method, not all support preferences, etc. not sure if there is a good way to abstract.

Perhaps better to just put an example at the docs 🤷
